### PR TITLE
reorder structs for better memory usage

### PR DESCRIPTION
**Pull request by florianl**
Signed-off-by: Florian Lehner <florian.lehner@elastic.co>

Reorder structs to reduce required memory.

### CODE REVIEW

File: audit.go [hunk 1]
```diff
@@ -404,8 +404,8 @@ func (c *AuditClient) SetBacklogWaitTime(waitTime int32, wm WaitMode) error {
 
 // RawAuditMessage is a raw audit message received from the kernel.
 type RawAuditMessage struct {
```
Comment (florianl) [inline audit.go:406]: struct with 16 pointer bytes could be 8

File: netlink.go [hunk 2]
```diff
@@ -59,13 +59,13 @@ type NetlinkParser func([]byte) ([]syscall.NetlinkMessage, error)
 
 // NetlinkClient is a generic client for sending and receiving netlink messages.
 type NetlinkClient struct {
```
Comment (florianl) [inline netlink.go:61]: struct with 88 pointer bytes could be 56

File: reassembler.go [hunk 3]
```diff
@@ -54,15 +54,16 @@ type Stream interface {
 // sequence number unless a late message is received that falls outside of the
 // sequences held in memory.
 type Reassembler struct {
```
Comment (florianl) [inline reassembler.go:56]: struct with 32 pointer bytes could be 24

File: reassembler.go [hunk 4]
```diff
@@ -197,12 +198,12 @@ func (e *event) IsExpired() bool {
 // Type - eventList
 
 type eventList struct {
```
Comment (florianl) [inline reassembler.go:200]: struct with 40 pointer bytes could be 16

**Review by andrewkroh — COMMENTED** [review]
It looks like auparse.AuditMessage could also be reordered. 
Do you want to fix that one too?

Screen Shot 2022-07-08 at 14 06 59


```
diff --git a/auparse/auparse.go b/auparse/auparse.go
index ff18f6a..cc80752 100644
--- a/auparse/auparse.go
+++ b/auparse/auparse.go
@@ -51,15 +51,14 @@ var (
 
 // AuditMessage represents a single audit message.
 type AuditMessage struct {
-       RecordType AuditMessageType // Record type from netlink header.
-       Timestamp  time.Time        // Timestamp parsed from payload in netlink message.
-       Sequence   uint32           // Sequence parsed from payload.
-       RawData    string           // Raw message as a string.
-
-       offset int               // offset is the index into RawData where the header ends and message begins.
-       data   map[string]string // The key value pairs parsed from the message.
-       tags   []string          // The keys associated with the event (e.g. the values set in rules with -F key=exec).
-       error  error             // Error that occurred while parsing.
+       Timestamp  time.Time         // Timestamp parsed from payload in netlink message.
+       error      error             // Error that occurred while parsing.
+       data       map[string]string // The key value pairs parsed from the message.
+       RawData    string            // Raw message as a string.
+       tags       []string          // The keys associated with the event (e.g. the values set in rules with -F key=exec).
+       offset     int               // offset is the index into RawData where the header ends and message begins.
+       Sequence   uint32            // Sequence parsed from payload.
+       RecordType AuditMessageType  // Record type from netlink header.
 }
```

File: reassembler.go [hunk 5]
```diff
@@ -197,12 +198,12 @@ func (e *event) IsExpired() bool {
...
-	sync.Mutex
-	seqs    sequenceNumSlice
 	events  map[sequenceNum]*event
-	lastSeq sequenceNum
+	seqs    sequenceNumSlice
 	maxSize int
 	timeout time.Duration
+	sync.Mutex
```
Comment (efd6) [inline reassembler.go:205]: The location of the mutex in the struct has human semantic burden; mutexes are conventionally placed above struct fields they protect in Go idiom.

**Comment by efd6** [comment]
I'd like to see what impact this has on time spent in `runtime.scanobject`.

File: reassembler.go (hunk 5 above)
Comment (florianl) [reply reassembler.go]: reverted this change.

**Comment by florianl** [comment]
> It looks like auparse.AuditMessage could also be reordered.
Do you want to fix that one too?

done :white_check_mark:

**Comment by efd6** [comment]
This is the output of the benchmarking that I did. Apart from very large cases (padding with 2^16 bytes) there is no difference in behaviours between front loaded pointers and rear loaded pointers. There is a significant difference between slab-allocated and element-wise allocation as expected and which I think is more likely the cause of the problem that this is intended to address.
 
<table class='benchstat '>
<tbody>
<tr><th><th>time/op
<tr><td>First1-16<td>7.32ms ± 7%
<tr><td>Last1-16<td>6.76ms ± 8%
<tr><td>First2-16<td>7.30ms ± 8%
<tr><td>Last2-16<td>6.85ms ± 7%
<tr><td>First4-16<td>7.85ms ± 9%
<tr><td>Last4-16<td>8.16ms ± 8%
<tr><td>First8-16<td>38.7ms ± 7%
<tr><td>Last8-16<td>39.0ms ± 4%
<tr><td>First16-16<td>607ms ±80%
<tr><td>Last16-16<td>541ms ± 3%
<tr><td>First1LL-16<td>45.0ms ± 4%
<tr><td>Last1LL-16<td>46.0ms ± 3%
<tr><td>First2LL-16<td>44.2ms ± 3%
<tr><td>Last2LL-16<td>45.3ms ± 2%
<tr><td>First4LL-16<td>47.2ms ± 4%
<tr><td>Last4LL-16<td>49.5ms ± 3%
<tr><td>First8LL-16<td>113ms ± 2%
<tr><td>Last8LL-16<td>165ms ± 5%
<tr><td>First16LL-16<td>709ms ± 2%
<tr><td>Last16LL-16<td>1.79s ± 5%
</tbody>
</table>

The details of the benchmarks:

There are a family of types (sub `N` with the numbers above)
```
type last1 struct {
	pad [1 << N]byte
	n   *lastN
}

type first1 struct {
	n   *firstN
	pad [1 << N]byte
}
```
and each was then benchmarked with the following for slab allocation (16 used only 1e5 due to memory constraint).
```
func BenchmarkFirstN(b *testing.B) {
	for i := 0; i < b.N; i++ {
		a := make([]firstN, 1e6)
		for j := range a {
			a[j].n = &a[(j+1)%len(a)]
		}
	}
}

func BenchmarkLastN(b *testing.B) {
	for i := 0; i < b.N; i++ {
		a := make([]lastN, 1e6)
		for j := range a {
			a[j].n = &a[(j+1)%len(a)]
		}
	}
}
```
and element-wise allocation
```
func BenchmarkFirstNLL(b *testing.B) {
	for i := 0; i < b.N; i++ {
		const n = 1e6
		h := &firstN{}
		orig := h
		for j := 1; j < n; j++ {
			h = &firstN{n: h}
		}
		h.n = orig
	}
}

func BenchmarkLastNLL(b *testing.B) {
	for i := 0; i < b.N; i++ {
		const n = 1e6
		h := &lastN{}
		orig := h
		for j := 1; j < n; j++ {
			h = &lastN{n: h}
		}
		h.n = orig
	}
}
```

**Comment by efd6** [comment]
I agree that if the changes made a difference that it would be significant given the proportion of the time spent in `scanobject`, but the change in ordering doesn't seem to as far as I can see. At least not for the sizes that we are dealing with.
```
$ go test -bench Last8$ -count 10 -cpuprofile last8.pprof
goos: darwin
goarch: amd64
pkg: scanobjord
cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
BenchmarkLast8-16    	      31	  37112835 ns/op
BenchmarkLast8-16    	      32	  37242561 ns/op
BenchmarkLast8-16    	      31	  38081059 ns/op
BenchmarkLast8-16    	      27	  37187969 ns/op
BenchmarkLast8-16    	      32	  37338076 ns/op
BenchmarkLast8-16    	      34	  35064717 ns/op
BenchmarkLast8-16    	      32	  35461664 ns/op
BenchmarkLast8-16    	      32	  37593525 ns/op
BenchmarkLast8-16    	      31	  38852988 ns/op
BenchmarkLast8-16    	      33	  35373119 ns/op
PASS
ok  	scanobjord	14.901s
$ go test -bench First8$ -count 10 -cpuprofile first8.pprof
goos: darwin
goarch: amd64
pkg: scanobjord
cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
BenchmarkFirst8-16    	      31	  35894025 ns/op
BenchmarkFirst8-16    	      28	  37102746 ns/op
BenchmarkFirst8-16    	      31	  36628196 ns/op
BenchmarkFirst8-16    	      33	  37609142 ns/op
BenchmarkFirst8-16    	      32	  35530509 ns/op
BenchmarkFirst8-16    	      32	  37586055 ns/op
BenchmarkFirst8-16    	      28	  37761853 ns/op
BenchmarkFirst8-16    	      32	  35432062 ns/op
BenchmarkFirst8-16    	      32	  37318492 ns/op
BenchmarkFirst8-16    	      31	  35174971 ns/op
PASS
ok  	scanobjord	14.623s
$ go tool pprof scanobjord.test last8.pprof  
File: scanobjord.test
Type: cpu
Time: Jul 11, 2022 at 7:00pm (ACST)
Duration: 14.69s, Total samples = 44.07s (299.96%)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top
Showing nodes accounting for 40560ms, 92.04% of 44070ms total
Dropped 87 nodes (cum <= 220.35ms)
Showing top 10 nodes out of 43
      flat  flat%   sum%        cum   cum%
   19240ms 43.66% 43.66%    23600ms 53.55%  runtime.scanobject
    4610ms 10.46% 54.12%     4610ms 10.46%  runtime.memmove
    4020ms  9.12% 63.24%     4570ms 10.37%  runtime.heapBitsSetType
    3050ms  6.92% 70.16%     3050ms  6.92%  runtime.memclrNoHeapPointers
    2710ms  6.15% 76.31%     3630ms  8.24%  runtime.gcWriteBarrier
    2330ms  5.29% 81.60%     2330ms  5.29%  runtime.heapBits.bits (inline)
    1850ms  4.20% 85.80%     1960ms  4.45%  runtime.heapBits.next (inline)
    1270ms  2.88% 88.68%    12570ms 28.52%  scanobjord.BenchmarkLast8
     850ms  1.93% 90.61%      860ms  1.95%  runtime.pthread_kill
     630ms  1.43% 92.04%      700ms  1.59%  runtime.pthread_cond_wait
(pprof) quit
$ go tool pprof scanobjord.test first8.pprof
File: scanobjord.test
Type: cpu
Time: Jul 11, 2022 at 7:01pm (ACST)
Duration: 14.50s, Total samples = 43.02s (296.69%)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top
Showing nodes accounting for 39640ms, 92.14% of 43020ms total
Dropped 72 nodes (cum <= 215.10ms)
Showing top 10 nodes out of 53
      flat  flat%   sum%        cum   cum%
   18720ms 43.51% 43.51%    23010ms 53.49%  runtime.scanobject
    4710ms 10.95% 54.46%     4710ms 10.95%  runtime.memmove
    4030ms  9.37% 63.83%     4510ms 10.48%  runtime.heapBitsSetType
    2710ms  6.30% 70.13%     3620ms  8.41%  runtime.gcWriteBarrier
    2690ms  6.25% 76.38%     2690ms  6.25%  runtime.memclrNoHeapPointers
    2430ms  5.65% 82.03%     2430ms  5.65%  runtime.heapBits.bits (inline)
    1730ms  4.02% 86.05%     1780ms  4.14%  runtime.heapBits.next (inline)
    1390ms  3.23% 89.28%    12290ms 28.57%  scanobjord.BenchmarkFirst8
     670ms  1.56% 90.84%      680ms  1.58%  runtime.pthread_kill
     560ms  1.30% 92.14%      640ms  1.49%  runtime.pthread_cond_wait
```

**Comment by florianl** [comment]
friendly ping @andrewkroh && @efd6

**Comment by florianl** [comment]
friendly ping @andrewkroh && @efd6

**Comment by florianl** [comment]
friendly ping @andrewkroh && @efd6 && @adriansr

**Comment by florianl** [comment]
friendly ping @andrewkroh && @efd6 && @adriansr

**Comment by andrewkroh** [comment]
> The problem I see with the posted benchmark is, that is focusing purely on the written benchmark.
> There are way more allocations happening in the scope of auditbeat and my goal is to reduce this.

It seems like we should be able to create a benchmark that allocates similarly to Auditbeat and run it for long enough that it demonstrates an improvement either through the benchmark itself or via some other observation like CPU profiles from the benchmark, execution traces, etc. I'm just looking for some evidence that we are getting a benefit here.

**Pull request closed without merging**